### PR TITLE
ENH: Refactor & Documentation in Decomposition module

### PR DESCRIPTION
**Pull request by KamalakerDadi**
- Made ```fit()``` usable from BaseDecomposition module for all decomposition estimators.
- This ```fit()``` can also be adapted to Parcellations object which is almost in ready in PR #1370 
- Added documentation for missing attributes in ```CanICA``` and ```DictLearning```.
- Added documentation for ```Returns``` in ```BaseDecomposition``` ```fit()```.

Can I have your opinions on this ?

### CODE REVIEW

**Comment by bthirion** [comment]
LGTM.

File: nilearn/decomposition/base.py [hunk 1]
```diff
@@ -345,7 +345,7 @@ def __init__(self, n_components=20,
         self.verbose = verbose
 
     def fit(self, imgs, y=None, confounds=None):
-        """Base fit for decomposition estimators : compute the embedded masker
+        """Compute the mask and the components across subjects
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/base.py]: Maybe "components" -> "PCA componets"

Comment (KamalakerDadi) [reply nilearn/decomposition/base.py]: I would not because this fit() is common to all estimators.

File: nilearn/decomposition/base.py [hunk 2]
```diff
@@ -375,6 +388,20 @@ def fit(self, imgs, y=None, confounds=None):
             self.masker_.fit()
         self.mask_img_ = self.masker_.mask_img_
 
+        # mask_and_reduce step is used only for decomposition estimators i.e.
+        # MultiPCA, CanICA and Dictionary Learning
+        if self.__class__.__name__ != 'BaseDecomposition':
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/base.py]: You don't need this if statement: BaseDecomposition is never used by end user. You can remove the if and simply insert the code that is below.

**Comment by GaelVaroquaux** [comment]
> I would not because this fit() is common to all estimators.

OK. Point taken.

File: nilearn/decomposition/base.py (hunk 2 above)
Comment (KamalakerDadi) [reply nilearn/decomposition/base.py]: tests will fail because BaseDecomposition has no _raw_fit method implemented.

**Comment by GaelVaroquaux** [comment]
> tests will fail because BaseDecomposition has no _raw_fit method implemented.

Do tests run BaseDecomposition.fit? They shouldn't

**Comment by KamalakerDadi** [comment]
>tests will fail because BaseDecomposition has no _raw_fit method implemented.
>Do tests run BaseDecomposition.fit? They shouldn't

Tests for BaseDecomposition module in nilearn/decomposition/tests/test_base.py

**Comment by GaelVaroquaux** [comment]
OK: here is a suggestion: implement a raw_fit method for
BaseDecomposition that does nothing (and do a comment in it to explain
why you do this).

File: nilearn/decomposition/canica.py [hunk 3]
```diff
@@ -178,6 +217,8 @@ def _unmix_components(self):
         for component in self.components_:
             if component.max() < -component.min():
                 component *= -1
+        if hasattr(self, "masker_"):
+            self.components_img_ = self.masker_.inverse_transform(self.components_)
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/canica.py]: This is great! Can you also propagate the change to the examples (and add a comment above the use of components_img_ exists only in 0.4.1 and after, and that users should unmask 'components_' for older versions).

File: nilearn/decomposition/dict_learning.py [hunk 4]
```diff
@@ -130,6 +146,27 @@ class DictLearning(BaseDecomposition, TransformerMixin):
     verbose: integer, optional
         Indicate the level of verbosity. By default, nothing is printed.
 
+    Attributes
+    ----------
+    `components_` : 2D numpy array (n_components x n-voxels)
+        Dictionary components extracted from the images.
+
+    `components_img_` : Nifti image
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/dict_learning.py]: 4D Nifti image

File: nilearn/decomposition/dict_learning.py [hunk 5]
```diff
@@ -130,6 +146,27 @@ class DictLearning(BaseDecomposition, TransformerMixin):
     verbose: integer, optional
         Indicate the level of verbosity. By default, nothing is printed.
 
+    Attributes
+    ----------
+    `components_` : 2D numpy array (n_components x n-voxels)
+        Dictionary components extracted from the images.
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/dict_learning.py]: Masked dictionary components....

File: nilearn/decomposition/dict_learning.py [hunk 6]
```diff
@@ -130,6 +146,27 @@ class DictLearning(BaseDecomposition, TransformerMixin):
...
+
+    `components_img_` : Nifti image
+        Extracted components unmasked to image.
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/dict_learning.py]: 4D image giving the extracted components...

File: nilearn/decomposition/tests/test_canica.py [hunk 7]
```diff
@@ -141,3 +143,43 @@ def test_threshold_bound():
...
+    canica.fit(data)
+
+
+def test_components_img():
+    data, mask_img, _, _ = _make_canica_test_data(n_subjects=3)
+    canica = CanICA(n_components=3, mask=mask_img)
+    canica.fit(data)
+    components_img = canica.components_img_
+    assert_true(isinstance(components_img, nibabel.Nifti1Image))
+    assert_not_equal(components_img, '')
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/tests/test_canica.py]: You should rather test the shape here.

File: nilearn/decomposition/tests/test_dict_learning.py [hunk 8]
```diff
@@ -74,3 +76,43 @@ def test_component_sign():
...
+    dict_learning.fit(data)
+
+
+def test_components_img():
+    data, mask_img, _, _ = _make_canica_test_data(n_subjects=3)
+    dict_learning = DictLearning(n_components=3, mask=mask_img)
+    dict_learning.fit(data)
+    components_img = dict_learning.components_img_
+    assert_true(isinstance(components_img, nibabel.Nifti1Image))
+    assert_not_equal(components_img, '')
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/tests/test_dict_learning.py]: Same comment here

File: nilearn/decomposition/tests/test_multi_pca.py [hunk 9]
```diff
@@ -66,7 +76,17 @@ def test_multi_pca():
...
+    assert_raises_regex(ValueError,
+                        'Need one or more Niimg-like objects as input, '
+                        'an empty list was given.',
+                        multi_pca.fit, [])
+    # Test passing masker arguments to estimator
+    multi_pca = MultiPCA(target_affine=affine,
+                         target_shape=shape[:3],
+                         n_components=3,
+                         mask_strategy='background')
+    multi_pca.fit(data)
```
Comment (GaelVaroquaux) [inline nilearn/decomposition/tests/test_multi_pca.py]: Can you add a components_img_ to MultiPCA and test it too. Thanks!

**Comment by GaelVaroquaux** [comment]
This is almost there: the general idea is good, only small comments.

**Comment by KamalakerDadi** [comment]
This needs rebase and adapting tests in decomposition estimators to attribute ```components_img_```.

**Comment by KamalakerDadi** [comment]
One thing I would like to confirm is whether to deprecate the attribute ```components_``` from next release version ? @bthirion @GaelVaroquaux 

Since we have added new attribute ```components_img_```, deprecating and removing ```components_``` in next two releases seems valid. And with this we would avoid duplicating returning same attributes. Moreover, simplifies the objects.

What do you think. Should I add ```.. deprecated:: 0.4.1``` at ```components_``` attribute ?

**Comment by GaelVaroquaux** [comment]
+1 for deprecation, @KamalakerDadi. Thanks!

**Comment by GaelVaroquaux** [comment]
Also, there is a failing PEP8

**Comment by GaelVaroquaux** [comment]
+1 for merge once the note on deprecation is added, and the PEP8 failure fixed.

**Comment by GaelVaroquaux** [comment]
Thank you very much. Merging!

**Comment by KamalakerDadi** [comment]
Thanks a lot for merging.

**Pull request merged**